[receiver/awscloudwatchmetricsreceiver] Add new receiver

lewis262626 opened the pull request:
**Description:** New AWS CloudWatch metrics receiver initial PR

**Link to tracking Issue:** #15667 

**Testing:** Unit tests for component config and factory

PR supersedes: #19218

lewis262626 committed:
Feat(receiver): Initial commit of receiver

Introduces the AWS CloudWatch metrics receiver based on issue #15667

lewis262626 commented:
cc @jpkrohling @kovrus

lewis262626 committed:
Chore(docs): Add changelog entry for receiver

Add chloggen entry for new metric receiver component

lewis262626 committed:
fix: Add link to changelog entry

lewis262626 committed:
chore: update github issue templates

atoulme commented on .github/ISSUE_TEMPLATE/feature_request.yaml:
```diff
@@ -1,4 +1,4 @@
-name: Feature request
+eame: Feature request
```
```suggestion
name: Feature request
```

lewis262626 committed:
Update .github/ISSUE_TEMPLATE/feature_request.yaml

Co-authored-by: Antoine Toulme <antoine@toulme.name>

kovrus commented on receiver/awscloudwatchmetricsreceiver/config.go:
```diff
@@ -0,0 +1,179 @@
+
+	// https://docs.aws.amazon.com/cli/latest/reference/cloudwatch/get-metric-data.html
+	// GetMetricData supports up to 500 metrics per API call
+	errTooManyMetrics = errors.New("too many metrics defined")
```
Maybe we shall add the information about the limit to the error message. You've done that for `errTooManyDimensions`, so let's add it here too.

lewis262626 replied:
Fixed in [8ab9607](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/19429/commits/8ab9607cd8a44fe14c0e9f55646914d57d3ee8fb)

kovrus reviewed (approved):
LGTM

lewis262626 committed:
Chore(deps): Bump receiver deps

lewis262626 committed:
Fix(errors): Improve tooManyMetrics error message

As per GH comment, include the number of metrics you can define in the
error message.

Lowercase debug messages for start and shutdown receiver method

lewis262626 commented:
Unsure why CI is failing. Logs are saying it's failing on a Jaeger test 🤔

lewis262626 committed:
Chore: run make goporto

atoulme commented:
It’s failing on checks too, please take a look

lewis262626 commented:
@atoulme I fixed the issue with `make goporto`, but the `jaeger` issue seems to be nothing related to my commits. The logs say something is already listening on that port:

https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/4387063396/jobs/7685532139#step:9:520

atoulme commented:
It's a flaky test: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9113

lewis262626 commented:
> It's a flaky test: #9113

Ok, hopefully a re-run fixes it

lewis262626 committed:
Chore: make gendependabot

atoulme commented:
Please add the new module to versions.yaml with `Using versioning file /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/versions.yaml`

lewis262626 committed:
Chore: add module to `versions.yaml`

lewis262626 commented:
> Please add the new module to versions.yaml with `Using versioning file /home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/versions.yaml`

Cheers, fixed.

lewis262626 committed:
chore: dp templates

lewis262626 commented:
Can someone with write permissions merge this PR please?

matej-g reviewed (commented):
Thanks for this @lewis262626!

I left a few questions / suggestions. It seems like the receiver does not implement any CloudWatch API polling yet? Is this meant to provide just bare bones for the receiver? 

(Sorry if that's just my unfamiliarity with process, not sure what is the required minimum viable functionality for a receiver to be merged. I was expecting an initial version to already be capable of receiving metrics).

lewis262626 commented on receiver/awscloudwatchmetricsreceiver/receiver.go:
```diff
@@ -0,0 +1,78 @@
+	doneChan      chan bool
+}
+
+func newMetricsRceiver(cfg *Config, logger *zap.Logger, consumer consumer.Metrics) *metricReceiver {
```
Will fix, cheers

lewis262626 replied:
Fixed in 92c82032c7

lewis262626 commented on receiver/awscloudwatchmetricsreceiver/README.md:
```diff
@@ -0,0 +1,145 @@
+
+## Getting Started
+
+This receiver uses the [AWS SDK](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html) as mode of authentication, which includes [Profile](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html) and [IMDS](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html) authentication for EC2 instances.
```
Typo, should be v2. Thanks!

lewis262626 replied:
Fixed in df9aefd825d1267621fff213c0ef700c556ffc78

lewis262626 commented on receiver/awscloudwatchmetricsreceiver/config.go:
```diff
@@ -0,0 +1,179 @@
+	}
+	for _, name := range metricsNames {
+		if name.Namespace == "" {
+			return errNoMetricsConfigured
```
Makes sense, will add an extra error and update tests

lewis262626 replied:
Fixed in df9aefd825d1267621fff213c0ef700c556ffc78

lewis262626 commented on receiver/awscloudwatchmetricsreceiver/config.go:
```diff
@@ -0,0 +1,179 @@
+)
+
+var (
+	defaultPollInterval = time.Minute
```
It depends, you can get get high precision AWS metrics that have a 1s period. I think for AWS metrics it's 5m, so let's do that. Cheers

lewis262626 replied:
Fixed in 5bcaf6eed472b209affdfad37bfd1914508c5126

lewis262626 commented on receiver/awscloudwatchmetricsreceiver/config.go:
```diff
@@ -0,0 +1,179 @@
+
+	// https://docs.aws.amazon.com/cli/latest/reference/cloudwatch/get-metric-data.html
+	// GetMetricData supports up to 500 metrics per API call
+	errTooManyMetrics = errors.New("you cannot define more than 500 metrics")
```
I think this can be removed as I have implemented the split logic already

lewis262626 replied:
Fixed in df9aefd825d1267621fff213c0ef700c556ffc78

matej-g commented:
@lewis262626 thanks for the explanation, let's get this in! ❤️

codeboten reviewed (commented):
Please resolve/address the remaining comments and we can get this merged!

jpkrohling reviewed (commented):
dependabot conflict solved, but apparently, there are a few unhandled comments. I'll go over this once the comments are taken care of.

lewis262626 commented:
> @lewis262626 thanks for the explanation, let's get this in! ❤️ 

I have some free time later this week. Will resolve the outstanding comments and rebase

lewis262626 committed:
Chore: Update based on PR comments

lewis262626 committed:
Chore: update tests based on PR comments

lewis262626 committed:
fix(spelling): Fix function spelling

TylerHelmuth commented:
@jpkrohling please give final review as the sponsor.

atoulme commented on .github/CODEOWNERS:
```diff
@@ -159,6 +159,7 @@ receiver/activedirectorydsreceiver/                      @open-telemetry/collect
 receiver/aerospikereceiver/                              @open-telemetry/collector-contrib-approvers @djaglowski @antonblock
 receiver/apachereceiver/                                 @open-telemetry/collector-contrib-approvers @djaglowski
 receiver/awscloudwatchreceiver/                          @open-telemetry/collector-contrib-approvers @djaglowski @schmikei
+receiver/awscloudwatchmetricsreceiver/                   @open-telemetry/collector-contrib-approvers @lewis262626 @jpkrohling @kovrus
```
```suggestion
receiver/awscloudwatchmetricsreceiver/                   @open-telemetry/collector-contrib-approvers @jpkrohling @kovrus
```

Given that you're not a member of OpenTelemetry yet, we cannot list you as codeowner. Please see https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/20868 as background.

An easy way to go about this is to remove your name for the time being, merge this in and use me and the sponsor of the component @jpkrohling as sponsors to get you added as member of the OpenTelemetry organization.

atoulme reviewed (commented):
Please make sure to do the following:
* [x] rebase with latest main
* [x] run `make gendependabot`
* [x] update all version numbers to 0.77.0

That should be enough to get you ready for merge.

lewis262626 committed:
Update .github/CODEOWNERS

Co-authored-by: Antoine Toulme <antoine@toulme.name>

lewis262626 committed:
Chore: make gendependabot

lewis262626 committed:
Chore: bump deps

lewis262626 commented:
@atoulme That should be all done. Appreciate your help :)

Great to get this approved and merged in this week

matej-g reviewed (approved):
Let's get this in 💪 , thanks @lewis262626!

lewis262626 committed:
Fix(docs): update aws cli link

lewis262626 committed:
Chore: make gotidy

codeboten commented on receiver/awscloudwatchmetricsreceiver/README.md:
```diff
@@ -0,0 +1,144 @@
+| ------------------------ | ------------- |
+| Stability                | [development] |
+| Supported pipeline types | metrics       |
+| Distributions            | [contrib]     |
```
Doesn't have to be in this PR, but please follow up with an update to auto-generate the status information

jpkrohling commented on receiver/awscloudwatchmetricsreceiver/config_test.go:
```diff
@@ -0,0 +1,236 @@
+						Period:         time.Second * 60,
+						AwsAggregation: "Sum",
+						Dimensions: []MetricDimensionsConfig{{
+							Name:  "Test",
```
Is there a real value that would have meaning to users? Tests are documentation as well, using real-world examples to help people understand the code under test.

lewis262626 replied:
Agreed. I'll fix this

lewis262626 replied:
I'll fix the remaining issues and rebase tomorrow

lewis262626 committed:
Update receiver/awscloudwatchmetricsreceiver/config_test.go

Co-authored-by: Juraci Paixão Kröhling <juraci.github@kroehling.de>

lewis262626 commented on receiver/awscloudwatchmetricsreceiver/config.go:
```diff
@@ -0,0 +1,173 @@
+	if cfg.Metrics == nil {
+		return errNoMetricsConfigured
+	}
+	return cfg.validateNamedConfig()
```
I can see both angles tbh, probably worth a follow-up issue to see if some advice can go in to the `CONTRIBUTING.md` or the opentelemetry docs

lewis262626 committed:
Update receiver/awscloudwatchmetricsreceiver/config_test.go

Co-authored-by: Juraci Paixão Kröhling <juraci.github@kroehling.de>

lewis262626 commented on receiver/awscloudwatchmetricsreceiver/config_test.go:
```diff
@@ -0,0 +1,236 @@
+						Period:         time.Second * 60,
+						AwsAggregation: "Sum",
+						Dimensions: []MetricDimensionsConfig{{
+							Name:  ":",
```
```suggestion
							Name:  ":P99",
```

lewis262626 committed:
Update receiver/awscloudwatchmetricsreceiver/config_test.go

lewis262626 commented on receiver/awscloudwatchmetricsreceiver/README.md:
```diff
@@ -0,0 +1,144 @@
+
+## AWS Costs
+
+This receiver uses the [GetMetricData](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_GetMetricData.html) API call, this call is *not* in the AWS free tier. Please refer to [Amazon's pricing](https://aws.amazon.com/cloudwatch/pricing/) for further information about expected costs. For `us-east-1`, the current pricing is $0.01 per 1,000 metrics requested as of February 2023.
```
I'm just following what the AWS docs say. I can make it explicit that it's the number of metrics you define in the receiver configuration, and not how many API calls the receiver makes.

lewis262626 replied:
I think it's worth just deleting that sentence. The pricing changes fairly frequently and varies between regions. It's the user's decision to weigh up pricing etc. None of the other receivers such as Splunk etc have pricing. 

Large customers could also have discounted pricing etc.

I can follow up with an issue to see if a custom collector metric around number of api calls and number of requested metrics would be useful?

lewis262626 committed:
chore: test formatting

lewis262626 commented on receiver/awscloudwatchmetricsreceiver/config_test.go:
```diff
@@ -0,0 +1,236 @@
+						AwsAggregation: "TS99",
+					},
+						{
+							Namespace:      "AWS/EC2",
```
Great spot. Fixed

jpkrohling replied:
Is this fixed? I still see something odd here. I won't block the PR because of this, but please double-check on a follow-up PR.

lewis262626 committed:
docs: delete pricing sentence

lewis262626 committed:
Chore: update tests with proper examples

lewis262626 committed:
chore: make gendependabot

lewis262626 commented:
Should all be fixed. Appreciate a final review @jpkrohling

dmitryax commented:
@lewis262626 please take a look at the failing linter

lewis262626 committed:
chore: make addlicense

lewis262626 commented:
> @lewis262626 please take a look at the failing linter

Fixed the linter issue
